### PR TITLE
Topology optimization

### PR DESCRIPTION
**Pull request by brendankeith**
This PR presents an MFEM example of topology optimization. This addresses a user request made at the first MFEM workshop.

The objective is to minimize the compliance
```math
\text{minimize} \int_\Omega f \cdot u\, \mathrm{d}x \text{~~over~~} u \in [H^1(\Omega)]^2 \text{~~and~~} \rho \in L^1(\Omega)
```
subject to
```math
-\mathrm{Div}(r(\tilde{\rho})\mathsf{C}\varepsilon(u)) = f  \quad \text{in~} \Omega ~+ ~ \text{BCs}
```
```math
-\epsilon^2 \Delta \tilde{\rho} + \tilde{\rho} = \rho  \quad \text{in~} \Omega ~ + ~ \text{Neumann BCs}
```
```math
0 \leq \rho \leq 1 \quad \text{in~} \Omega
\qquad
\text{and}
\qquad
\int_\Omega \rho\, \mathrm{d}x = \theta\, \mathrm{vol}(\Omega)
```

Here, $r(\tilde{\rho}) = \rho_0 + \tilde{\rho}^3 (1 - \rho_0)$ is the solid isotropic material penalization (SIMP) law, $\mathsf{C}$ is the elasticity tensor for an isotropic linearly elastic material, $\epsilon > 0$ is the design length scale, and $0 < θ < 1$ is the volume fraction.

The method used here is based on entropic mirror descent [1] and was first proposed in [2].

The code does not rely on external software packages (e.g., IPOPT, PETSc).

[1] Beck, A., & Teboulle, M. (2003). Mirror descent and nonlinear projected subgradient methods for convex optimization. Operations Research Letters, 31(3), 167-175.
[2] Keith, B. and Surowiec, T. (2023) Proximal Galerkin: A structure-preserving finite element method for pointwise bound constraints. arXiv:2307.12444 [math.NA].

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3400](https://github.com/mfem/mfem/pull/3400) | @brendankeith | @v-dobrev | @bslazarov + @psocratis | 1/16/23 | 9/14/23 | 9/25/23 | |

<details>
<summary>PR Checklist</summary>
    
- [x] Code builds.
- [x] Code passes `make style`.
- [x] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [x] All sample runs at the top of the example source file work.
    - [x] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [x] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [x] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by bslazarov** [comment]
The cmake file in the examples directory has to be updated.

File: examples/ex35.cpp [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+
+   // 6. Set-up the physics solver.
+   int maxat = mesh.bdr_attributes.Max();
+   Array<int> ess_bdr(maxat);
+   ess_bdr = 0;
+   ess_bdr[0] = 1;
+   ConstantCoefficient one(1.0);
+   ConstantCoefficient lambda_cf(lambda);
+   ConstantCoefficient mu_cf(mu);
+   LinearElasticitySolver * ElasticitySolver = new LinearElasticitySolver();
```
Comment (bslazarov) [inline examples/ex35.cpp:286]: the solver is not deleted!!!

File: examples/ex35.cpp [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+   Vector center(2); center(0) = 2.9; center(1) = 0.5;
+   Vector force(2); force(0) = 0.0; force(1) = -1.0;
+   double r = 0.05;
+   VolumeForceCoefficient vforce_cf(r,center,force);
+   ElasticitySolver->SetRHSCoefficient(&vforce_cf);
+   ElasticitySolver->SetEssentialBoundary(ess_bdr);
+
+   // 7. Set-up the filter solver.
+   ConstantCoefficient eps2_cf(epsilon*epsilon);
+   DiffusionSolver * FilterSolver = new DiffusionSolver();
```
Comment (bslazarov) [inline examples/ex35.cpp:299]: The solver is not deleted!!!

File: examples/ex35.cpp [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+         paraview_dc.SetTime((double)k);
+         paraview_dc.Save();
+      }
+
+      if (norm_reduced_gradient < tol)
+      {
+         break;
+      }
+   }
+
```
Comment (bslazarov) [inline examples/ex35.cpp:440]: delete the solvers

File: examples/ex35p.cpp [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+
+   // 6. Set-up the physics solver.
+   int maxat = pmesh.bdr_attributes.Max();
+   Array<int> ess_bdr(maxat);
+   ess_bdr = 0;
+   ess_bdr[0] = 1;
+   ConstantCoefficient one(1.0);
+   ConstantCoefficient lambda_cf(lambda);
+   ConstantCoefficient mu_cf(mu);
+   LinearElasticitySolver * ElasticitySolver = new LinearElasticitySolver();
```
Comment (bslazarov) [inline examples/ex35p.cpp:308]: delete the solver

File: examples/ex35p.cpp [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+   Vector center(2); center(0) = 2.9; center(1) = 0.5;
+   Vector force(2); force(0) = 0.0; force(1) = -1.0;
+   double r = 0.05;
+   VolumeForceCoefficient vforce_cf(r,center,force);
+   ElasticitySolver->SetRHSCoefficient(&vforce_cf);
+   ElasticitySolver->SetEssentialBoundary(ess_bdr);
+
+   // 7. Set-up the filter solver.
+   ConstantCoefficient eps2_cf(epsilon*epsilon);
+   DiffusionSolver * FilterSolver = new DiffusionSolver();
```
Comment (bslazarov) [inline examples/ex35p.cpp:321]: delete the solver at the end

File: examples/ex35p.cpp [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+         paraview_dc.SetTime((double)k);
+         paraview_dc.Save();
+      }
+
+      if (norm_reduced_gradient < tol)
+      {
+         break;
+      }
+   }
+
```
Comment (bslazarov) [inline examples/ex35p.cpp:472]: delete the elasticity and the diffusion solvers

File: examples/ex35p.cpp [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+      FilterSolver->Solve();
+      w_filter = *FilterSolver->GetFEMSolution();
+
+      // Step 4 - Compute gradient
+      // Solve G = M⁻¹w̃
+      GridFunctionCoefficient w_cf(&w_filter);
+      ParLinearForm w_rhs(&control_fes);
+      w_rhs.AddDomainIntegrator(new DomainLFIntegrator(w_cf));
+      w_rhs.Assemble();
+      M.Mult(w_rhs,grad);
```
Comment (bslazarov) [inline examples/ex35p.cpp:423]: Isn't that an approximate solution?

**Review by bslazarov — CHANGES_REQUESTED** [review]

File: examples/ex35.cpp [hunk 8]
```diff
@@ -0,0 +1,442 @@
+//                              MFEM Example 35
+//
+//
+// Compile with: make ex35
+//
+// Sample runs:
```
[the remark this answers is not in the document]
Comment (brendankeith) [reply examples/ex35.cpp:6]: Added comment about this

File: examples/ex35.hpp [hunk 9]
```diff
@@ -0,0 +1,690 @@
...
+   {
+      if (mesh->bdr_attributes.Size())
+      {
+         ess_bdr.SetSize(mesh->bdr_attributes.Max());
+         ess_bdr = 1;
+      }
+   }
+}
+
+void DiffusionSolver::Solve()
```
[the remark this answers is not in the document]
Comment (brendankeith) [reply examples/ex35.hpp:363]: Ah, yes, this was done so we could easily use the diffusion solver for thermal compliance problems. I agree that it is not optimally efficient the way it is. Still, the dominant cost lies in the elasticity solver, whose stiffness matrix does need to be updated at every iteration. So the sake of code simplicity, I would like to keep the diffusion solver as it is.

File: examples/ex35.hpp [hunk 10]
```diff
@@ -0,0 +1,690 @@
...
+{
+private:
+   Mesh * mesh = nullptr;
+   int order = 1;
+   Coefficient * diffcf = nullptr;
+   Coefficient * masscf = nullptr;
+   Coefficient * rhscf = nullptr;
+   Coefficient * essbdr_cf = nullptr;
+   Coefficient * neumann_cf = nullptr;
+   VectorCoefficient * gradient_cf = nullptr;
```
[the remark this answers is not in the document]
Comment (brendankeith) [reply examples/ex35.hpp:165]: Discussed this with you offline. It is useful for manufactured solutions

**Review by bslazarov — APPROVED** [review]

**Comment by dohyun-cse** [comment]
I just updated this branch with the following changes:

1. Set $rho_h=\text{sigmoid}(\psi_h)$ with $\psi_h\in V_h$.
In the code, this is done by defining `GridFunction psi(&control_fes)` and `MappedGridFunctionCoefficient rho(psi, sigmoid)`.
Here, `MappedGridFunctionCoefficient` is a child class of `GridFunctionCoefficient` returning `fun(gf->Eval(T, ip)`. In this example, `fun=sigmoid`.

2. Update `projit` with a more mathematically correct projection `psi = psi + c` where $c\in\mathbb{R}$ satisfies $\int_\Omega \text{sigmoid}(\psi + c) = \theta \text{vol}(\Omega)$.

In this way, we do not have to use $\text{sigmoid}^{-1}$ in the update procedure and can preserve $0\leq \rho \leq 1$ exactly. Also, we do not need `BasisType::Positive` anymore.
To prevent `psi` from diverging, i.e., $\rho \rightarrow 0$ or $\rho \rightarrow 1$, we apply `clip` after each mirror descent update.

**Comment by brendankeith** [comment]
> I just updated this branch with the following changes:
> 
> 1. Set ρh=sigmoid(ψh) with ψh∈Vh.
>    In the code, this is done by defining `GridFunction psi(&control_fes)` and `MappedGridFunctionCoefficient rho(psi, sigmoid)`.
>    Here, `MappedGridFunctionCoefficient` is a child class of `GridFunctionCoefficient` returning `fun(gf->Eval(T, ip)`. In this example, `fun=sigmoid`.
> 2. Update `projit` with a more mathematically correct projection `psi = psi + c` where c∈R satisfies ∫Ωsigmoid(ψ+c)=θvol(Ω).
> 
> In this way, we do not have to use sigmoid−1 in the update procedure and can preserve 0≤ρ≤1 exactly. Also, we do not need `BasisType::Positive` anymore. To prevent `psi` from diverging, i.e., ρ→0 or ρ→1, we apply `clip` after each mirror descent update.

Thank you, @dohyun-cse !!

**Comment by v-dobrev** [comment]
@bslazarov, can you please take another look after the latest changes?

**Comment by psocratis** [comment]
@brendankeith , I am a bit confused. It seems that this branch contains also the changes from #3398. Should we close the other one and review just his one?

**Comment by brendankeith** [comment]
> @brendankeith , I am a bit confused. It seems that this branch contains also the changes from #3398. Should we close the other one and review just his one?

The plan is to review and finish the review of https://github.com/mfem/mfem/pull/3398 and then merge that PR into this one. The only files to review in the topology optimization PR pertain to example 35, which make sense to review separately from the ex34 files.

File: examples/ex37.cpp [hunk 11]
```diff
@@ -0,0 +1,472 @@
...
+   int  visport   = 19916;
+   socketstream sout_r;
+   if (visualization)
+   {
+      sout_r.open(vishost, visport);
+      sout_r.precision(8);
+   }
+
+   rho_gf.ProjectCoefficient(rho);
+   mfem::ParaViewDataCollection paraview_dc("ex37", &mesh);
```
Comment (psocratis) [inline examples/ex37.cpp:386]: maybe ParView output should be added as a user option

File: examples/ex37p.cpp [hunk 12]
```diff
@@ -0,0 +1,506 @@
...
+   char vishost[] = "localhost";
+   int  visport   = 19916;
+   socketstream sout_r;
+   if (visualization)
+   {
+      sout_r.open(vishost, visport);
+      sout_r.precision(8);
+   }
+
+   rho_gf.ProjectCoefficient(rho);
```
Comment (psocratis) [inline examples/ex37p.cpp:412]: maybe ParView output should be added as a user option

**Review by psocratis — APPROVED** [review]
Very nice example!

**Review by bslazarov — APPROVED** [review]

File: examples/ex37.hpp [hunk 13]
```diff
@@ -0,0 +1,786 @@
...
+   }
+   else
+   {
+      MFEM_ABORT("Wrong code path. Call GetFEMSolution");
+      return nullptr;
+   }
+}
+#endif
+
+void DiffusionSolver::ResetFEM()
```
Comment (psocratis) [inline examples/ex37.hpp:580]: @brendankeith, it seems that there is no need for this function anymore. It's not used anywhere else but the destructor. Same for the `LinearElasticitySolver`

Comment (brendankeith) [reply examples/ex37.hpp]: Addressed

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
I have one request: is there a way to make the test runs of `ex37` and `ex37p` (for both GNU make and CMake) faster? Maybe use a coarser mesh or just make a few optimization iterations? Currently, the test runs of other example take a few seconds and for `ex37`/`ex37p` they take roughly 2min/1min.

**Comment by brendankeith** [comment]
> I have one request: is there a way to make the test runs of `ex37` and `ex37p` (for both GNU make and CMake) faster? Maybe use a coarser mesh or just make a few optimization iterations? Currently, the test runs of other example take a few seconds and for `ex37`/`ex37p` they take roughly 2min/1min.

@v-dobrev:

Reducing the number of iterations might not make sense. If we used much fewer iterations in any of the runs, we wouldn't resolve much of the optimal design.

I could easily reduce the number of test runs. However, there may be a pedagogical drawback to doing so.

As a potential compromise, is it possible to include additional test runs that aren't checked by the CI?

**Comment by v-dobrev** [comment]
When running `make test` (or `ctest` with CMake) we don't necessarily need to run the examples so that the results makes sense from the point of view of a user. We mainly want to exercise the main code paths, so using a smaller number of iterations or a coarser mesh should not be a problem since our aim is not to fully solve the problem.

I agree that running the examples without parameters or the sample runs in the source files should make sense and fully solve the problem of interest. What I suggest is to run the tests of the examples with a set of parameters that is just used in `make test` (and `ctest`). Such tests can be added similar to this (in make): https://github.com/mfem/mfem/blob/664dfa2f625fbf3dcfee0b3b46d90da9f404a008/examples/makefile#L130-L133 and this (in CMake): https://github.com/mfem/mfem/blob/664dfa2f625fbf3dcfee0b3b46d90da9f404a008/examples/CMakeLists.txt#L104-L105

**Comment by brendankeith** [comment]
> When running `make test` (or `ctest` with CMake) we don't necessarily need to run the examples so that the results makes sense from the point of view of a user. We mainly want to exercise the main code paths, so using a smaller number of iterations or a coarser mesh should not be a problem since our aim is not to fully solve the problem.
> 
> I agree that running the examples without parameters or the sample runs in the source files should make sense and fully solve the problem of interest. What I suggest is to run the tests of the examples with a set of parameters that is just used in `make test` (and `ctest`). Such tests can be added similar to this (in make):
> 
> https://github.com/mfem/mfem/blob/664dfa2f625fbf3dcfee0b3b46d90da9f404a008/examples/makefile#L130-L133
> 
> and this (in CMake):
> https://github.com/mfem/mfem/blob/664dfa2f625fbf3dcfee0b3b46d90da9f404a008/examples/CMakeLists.txt#L104-L105

Thanks, @v-dobrev!! This is exactly what I was thinking of.

File: examples/ex37p.cpp [hunk 14]
```diff
@@ -0,0 +1,501 @@
+//                            MFEM Example 37 - Parallel Version
+//
+//
+// Compile with: make ex37p
+//
+// Sample runs:
+// mpirun -np 4 ex37p -alpha 10 -pv
+// mpirun -np 4 ex37p -lambda 0.1 -mu 0.1
+// mpirun -np 4 ex37p -o 2 -alpha 5.0 -mi 50 -vf 0.4 -ntol 1e-5
+// mpirun -np 4 ex37p -r 6 -o 2 -alpha 10.0 -epsilon 0.02 -mi 50 -ntol 1e-5
```
Comment (v-dobrev) [inline examples/ex37p.cpp:10]: I noticed in our nightly tests that there were no sample runs tested from `ex37p`, so I looked here: there should be at least 2 spaces before `mpirun` in order for these sample runs to be tested.
```suggestion
//    mpirun -np 4 ex37p -alpha 10 -pv
//    mpirun -np 4 ex37p -lambda 0.1 -mu 0.1
//    mpirun -np 4 ex37p -o 2 -alpha 5.0 -mi 50 -vf 0.4 -ntol 1e-5
//    mpirun -np 4 ex37p -r 6 -o 2 -alpha 10.0 -epsilon 0.02 -mi 50 -ntol 1e-5
```

Comment (brendankeith) [reply examples/ex37p.cpp]: I'm glad that you noticed this, @v-dobrev!

**Comment by v-dobrev** [comment]
Re-merged in `next` for testing.

**Pull request merged**